### ISSUE TITLE
too slow to be usable

### ISSUE THREAD

**Issue by plasticfist** [ZeliardM/homebridge-kasa-python#21]
**Describe The Bug:**
The delay between on/off button click until actual switch changes state is now 10 seconds to 40 seconds.  I thought I'd try the latest, I thought it was completely broken at first, switched from debug to stable, try clicking on/off again, but then after a light came on unexpectedly and did some more tests I realized it (works?) but now has a _very_ long action delay.
Note, the On/Off log message appear quickly, but the Kasa device only switches after considerable delay.

**To Reproduce:**
- have more than a few Kasa devices online (e.g. 95)
- load plugin stable (or debug), click on any accessory, wait, wait, no keep waiting...

**Expected behavior:**
click on button to turn on light, it should turn on as quickly as in the Kasa app (near instantaneous)

**Logs:**

```
[12/23/2024, 4:48:31 PM] [KasaPython] [Office light] Setting On to: true for Office light
[12/23/2024, 4:48:50 PM] [KasaPython] [Office light] Setting On to: false for Office light
[12/23/2024, 4:58:52 PM] [KasaPython] [Office light] Setting On to: true for Office light
[12/23/2024, 4:59:47 PM] [KasaPython] [Office light] Setting On to: false for Office light
```

**Plugin Config:**

```json
{
    "name": "KasaPython",
    "enableCredentials": false,
    "pollingInterval": 5,
    "discoveryPollingInterval": 300,
    "offlineInterval": 7,
    "waitTimeUpdate": 100,
    "platform": "KasaPython"
}
```

**Screenshots:**
NA

**Environment:**
```
* **Plugin Version**: v2.6.13
* **Homebridge Version**: v1.8.5
* **Node.js Version**: v22.12.0
* **NPM Version**: ?
* **Operating System**:  Linux version 6.8.12-1-pve (build@proxmox) (gcc (Debian 12.2.0-14) 12.2.0, GNU ld (GNU Binutils for Debian) 2.40) #1 SMP PREEMPT_DYNAMIC PMX 6.8.12-1 (2024-08-05T16:17Z)

homebridge-kasa-python v2.6.13, node v22.12.0, homebridge v1.8.5, api v2.7
```

**Comment by ZeliardM** [comment]
@plasticfist Are you able to try the v2.7.0-beta? It requires Python 3.11 or greater. I am pushing several updates for some functionality improvements and I'd like you to test it. Let me know.

**Comment by ZeliardM** [comment]
@plasticfist give v2.7.0-beta.10 a try and let me know.

**Comment by ZeliardM** [comment]
@plasticfist v2.7.0-beta.11 is out now, please try it and let me know if the response time is better for devices.

**Comment by plasticfist** [comment]
just a note, beta.10 was non-functional

beta.11 looks great so far, the devices I tried were near instant changes.  Only curiosity is a message in the log, see below, shows end of discovery, and two on/off accessory tests...  with message 

```
[12/26/2024, 4:07:58 AM] [KasaPython] Adding HomeKit device: [Main light] IOT.SMARTPLUGSWITCH [8006C4DDAD807BC95969B6E49409945F20683700] at host [10.0.1.97]
(node:1449237) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 periodicDeviceDiscoveryComplete listeners added to [EventEmitter]. MaxListeners is 10. Use emitter.setMaxListeners() to increase limit
(Use `node --trace-warnings ...` to show where the warning was created)
[12/26/2024, 4:08:35 AM] [KasaPython] [Kitchen light] Updating [Switch.On] on Kitchen light to false
[12/26/2024, 4:08:37 AM] [KasaPython] [Kitchen light] Updating [Switch.On] on Kitchen light to true
[12/26/2024, 4:08:46 AM] [KasaPython] [Dining room light] Updating [Switch.On] on Dining room light to true
[12/26/2024, 4:08:47 AM] [KasaPython] [Dining room light] Updating [Switch.On] on Dining room light to false
(node:1449237) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 periodicDeviceDiscoveryComplete listeners added to [EventEmitter]. MaxListeners is 10. Use emitter.setMaxListeners() to increase limit
```

**Comment by ZeliardM** [comment]
Ok, excellent. Yea, I cache the device connections from discovery and let the API stay connected to them for faster connectivity. Otherwise, it was having to reconnect to each device every time you want it to do something which can take time. Minimal impact on network from my testing. I'll look into the emitter messages, will add the requested logging for the error and then if you can provide another log after beta.12 is out that would be greatly appreciated.

**Comment by ZeliardM** [comment]
It's weird that beta.10 didn't work as the main changes were in beta.10 not 11

**Comment by ZeliardM** [comment]
Found a way for emitter issue as well and will dynamically increase the number of available emitters based on the number of devices created. That will be in beta.12 as well.

**Comment by ZeliardM** [comment]
@plasticfist try beta.12 to see if the eventlistener messages go away

**Comment by plasticfist** [comment]
still there

```
(node:1490440) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 periodicDeviceDiscoveryComplete listeners added to [EventEmitter]. MaxListeners is 10. Use emitter.setMaxListeners() to increase limit
(Use `node --trace-warnings ...` to show where the warning was created)

```

**Comment by plasticfist** [comment]
and thought you might be interested in this, after a little bit it fired off another discovery, and threw an error...  On the debug version, so this might be normal...

```
[12/27/2024, 3:23:34 PM] [KasaPython] Discovering devices...
[12/27/2024, 3:23:46 PM] [KasaPython] Discovered 31 devices
(node:1490440) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 periodicDeviceDiscoveryComplete listeners added to [EventEmitter]. MaxListeners is 10. Use emitter.setMaxListeners() to increase limit
[12/27/2024, 3:25:24 PM] [KasaPython] [Office light] Updating [Switch.On] on Office light to true
[12/27/2024, 3:28:04 PM] [KasaPython] Error during getSysInfo post request: AxiosError: socket hang up
    at Function.AxiosError.from (file:///var/lib/homebridge/node_modules/homebridge-kasa-python/node_modules/axios/lib/core/AxiosError.js:92:14)
    at RedirectableRequest.handleRequestError (file:///var/lib/homebridge/node_modules/homebridge-kasa-python/node_modules/axios/lib/adapters/http.js:620:25)
    at RedirectableRequest.emit (node:events:524:28)
    at ClientRequest.eventHandlers.<computed> (/var/lib/homebridge/node_modules/homebridge-kasa-python/node_modules/follow-redirects/index.js:49:24)
    at ClientRequest.emit (node:events:524:28)
    at emitErrorEvent (node:_http_client:104:11)
    at Socket.socketOnEnd (node:_http_client:536:5)
    at Socket.emit (node:events:536:35)
    at endReadableNT (node:internal/streams/readable:1698:12)
    at processTicksAndRejections (node:internal/process/task_queues:90:21)
    at Axios.request (file:///var/lib/homebridge/node_modules/homebridge-kasa-python/node_modules/axios/lib/core/Axios.js:45:41)
    at processTicksAndRejections (node:internal/process/task_queues:105:5)
    at DeviceManager.getSysInfo (file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/deviceManager.ts:192:24)
    at file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/homekitPowerStrip.ts:36:36 {
  code: 'ECONNRESET',
  config: {
    transitional: {
      silentJSONParsing: true,
      forcedJSONParsing: true,
      clarifyTimeoutError: false
    },
    adapter: [ 'xhr', 'http', 'fetch' ],
    transformRequest: [ [Function: transformRequest] ],
    transformResponse: [ [Function: transformResponse] ],
    timeout: 0,
    xsrfCookieName: 'XSRF-TOKEN',
    xsrfHeaderName: 'X-XSRF-TOKEN',
    maxContentLength: -1,
    maxBodyLength: -1,
    env: { FormData: [Function], Blob: [class Blob] },
    validateStatus: [Function: validateStatus],
    headers: Object [AxiosHeaders] {
      Accept: 'application/json, text/plain, */*',
      'Content-Type': 'application/json',
      'User-Agent': 'axios/1.7.9',
      'Content-Length': '164',
      'Accept-Encoding': 'gzip, compress, deflate, br'
    },
    method: 'post',
    url: 'http://127.0.0.1:38829/getSysInfo',
    data: '{"device_config":{"connection_type":{"device_family":"IOT.SMARTPLUGSWITCH","encryption_type":"XOR","https":false},"host":"10.0.1.46","timeout":5,"uses_http":false}}'
  },
  request: <ref *1> Writable {
    _events: {
      close: undefined,
      error: [Function: handleRequestError],
      prefinish: undefined,
      finish: undefined,
      drain: undefined,
      response: [Function: handleResponse],
      socket: [Function: handleRequestSocket]
    },
    _writableState: WritableState {
      highWaterMark: 65536,
      length: 0,
      corked: 0,
      onwrite: [Function: bound onwrite],
      writelen: 0,
      bufferedIndex: 0,
      pendingcb: 0,
      [Symbol(kState)]: 17580812,
      [Symbol(kBufferedValue)]: null
    },
    _maxListeners: undefined,
    _options: {
      maxRedirects: 21,
      maxBodyLength: Infinity,
      protocol: 'http:',
      path: '/getSysInfo',
      method: 'POST',
      headers: [Object: null prototype],
      agents: [Object],
      auth: undefined,
      family: undefined,
      beforeRedirect: [Function: dispatchBeforeRedirect],
      beforeRedirects: [Object],
      hostname: '127.0.0.1',
      port: '38829',
      agent: undefined,
      nativeProtocols: [Object],
      pathname: '/getSysInfo'
    },
    _ended: true,
    _ending: true,
    _redirectCount: 0,
    _redirects: [],
    _requestBodyLength: 164,
    _requestBodyBuffers: [ [Object] ],
    _eventsCount: 3,
    _onNativeResponse: [Function (anonymous)],
    _currentRequest: ClientRequest {
      _events: [Object: null prototype],
      _eventsCount: 7,
      _maxListeners: undefined,
      outputData: [],
      outputSize: 0,
      writable: true,
      destroyed: false,
      _last: false,
      chunkedEncoding: false,
      shouldKeepAlive: true,
      maxRequestsOnConnectionReached: false,
      _defaultKeepAlive: true,
      useChunkedEncodingByDefault: true,
      sendDate: false,
      _removedConnection: false,
      _removedContLen: false,
      _removedTE: false,
      strictContentLength: false,
      _contentLength: '164',
      _hasBody: true,
      _trailer: '',
      finished: true,
      _headerSent: true,
      _closed: false,
      _header: 'POST /getSysInfo HTTP/1.1\r\n' +
        'Accept: application/json, text/plain, */*\r\n' +
        'Content-Type: application/json\r\n' +
        'User-Agent: axios/1.7.9\r\n' +
        'Content-Length: 164\r\n' +
        'Accept-Encoding: gzip, compress, deflate, br\r\n' +
        'Host: 127.0.0.1:38829\r\n' +
        'Connection: keep-alive\r\n' +
        '\r\n',
      _keepAliveTimeout: 0,
      _onPendingData: [Function: nop],
      agent: [Agent],
      socketPath: undefined,
      method: 'POST',
      maxHeaderSize: undefined,
      insecureHTTPParser: undefined,
      joinDuplicateHeaders: undefined,
      path: '/getSysInfo',
      _ended: false,
      res: null,
      aborted: false,
      timeoutCb: [Function: emitRequestTimeout],
      upgradeOrConnect: false,
      parser: null,
      maxHeadersCount: null,
      reusedSocket: true,
      host: '127.0.0.1',
      protocol: 'http:',
      _redirectable: [Circular *1],
      [Symbol(shapeMode)]: false,
      [Symbol(kCapture)]: false,
      [Symbol(kBytesWritten)]: 0,
      [Symbol(kNeedDrain)]: false,
      [Symbol(corked)]: 0,
      [Symbol(kChunkedBuffer)]: [],
      [Symbol(kChunkedLength)]: 0,
      [Symbol(kSocket)]: [Socket],
      [Symbol(kOutHeaders)]: [Object: null prototype],
      [Symbol(errored)]: null,
      [Symbol(kHighWaterMark)]: 65536,
      [Symbol(kRejectNonStandardBodyWrites)]: false,
      [Symbol(kUniqueHeaders)]: null
    },
    _currentUrl: 'http://127.0.0.1:38829/getSysInfo',
    [Symbol(shapeMode)]: true,
    [Symbol(kCapture)]: false
  },
  cause: Error: socket hang up
      at Socket.socketOnEnd (node:_http_client:536:25)
      at Socket.emit (node:events:536:35)
      at endReadableNT (node:internal/streams/readable:1698:12)
      at processTicksAndRejections (node:internal/process/task_queues:90:21) {
    code: 'ECONNRESET'
  }
}
[12/27/2024, 3:28:04 PM] [KasaPython] [Smart Plug 6A32] Error updating device state: TypeError: Cannot read properties of undefined (reading 'alias')
    at file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/homekitPowerStrip.ts:37:81
    at processTicksAndRejections (node:internal/process/task_queues:105:5)
```

**Comment by plasticfist** [comment]
After about half the discovery events, it throws exactly one error like the above, each time it has been a different device mentioned in the error message.

Also periodically seeing some messages like below, not the same device each time.   Note, I've never had a connection issue with the example device below through the Kasa app (this switch is ~20ft from a unifi enterprise wireless access point). 

```
[12/27/2024, 3:53:48 PM] [KasaPython] Discovered 33 devices
(node:1490440) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 periodicDeviceDiscoveryComplete listeners added to [EventEmitter]. MaxListeners is 10. Use emitter.setMaxListeners() to increase limit
[12/27/2024, 3:58:34 PM] [KasaPython] Discovering devices...
[12/27/2024, 3:58:46 PM] [KasaPython] Discovered 32 devices
(node:1490440) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 periodicDeviceDiscoveryComplete listeners added to [EventEmitter]. MaxListeners is 10. Use emitter.setMaxListeners() to increase limit
[12/27/2024, 3:59:55 PM] [KasaPython] [Porch light] Device is offline or platform is shutting down, cannot set value for characteristic On
[12/27/2024, 3:59:55 PM] [KasaPython] [Porch light] Device is offline or platform is shutting down, cannot set value for characteristic On
```

**Comment by ZeliardM** [comment]
@plasticfist understood, I'll look into it and let you know.

**Comment by ZeliardM** [comment]
@plasticfist give beta.14 a shot and let me know

**Comment by plasticfist** [comment]
initial discover a short pause and then this 

```
[12/28/2024, 1:42:21 AM] [KasaPython] Adding HomeKit device: [Main light] IOT.SMARTPLUGSWITCH [8006C4DDAD807BC95969B6E49409945F20683700] at host [10.0.1.97]
[12/28/2024, 1:42:59 AM] [KasaPython] [Theater light] Updating [Lightbulb.On] on Theater light to true
(node:33179) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 periodicDeviceDiscoveryComplete listeners added to [EventEmitter]. MaxListeners is 10. Use emitter.setMaxListeners() to increase limit
(Use `node --trace-warnings ...` to show where the warning was created)
[12/28/2024, 1:43:46 AM] [KasaPython] Exception during getSysInfo post request: Unable to query the device 10.0.1.144:9999: [Errno 104] Connection reset by peer
[12/28/2024, 1:43:46 AM] [KasaPython] [Power Strip 0324] Error updating device state: TypeError: Cannot read properties of undefined (reading 'alias')
    at file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/homekitPowerStrip.ts:37:81
    at processTicksAndRejections (node:internal/process/task_queues:105:5)
[12/28/2024, 1:45:01 AM] [KasaPython] Exception during getSysInfo post request: Unable to query the device 10.0.1.27:9999: [Errno 104] Connection reset by peer
[12/28/2024, 1:45:01 AM] [KasaPython] [Power Strip 9379] Error updating device state: TypeError: Cannot read properties of undefined (reading 'alias')
    at file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/homekitPowerStrip.ts:37:81
    at processTicksAndRejections (node:internal/process/task_queues:105:5)
[12/28/2024, 1:45:36 AM] [KasaPython] Exception during getSysInfo post request: Unable to query the device 10.0.1.142:9999: [Errno 104] Connection reset by peer
[12/28/2024, 1:45:36 AM] [KasaPython] [Power Strip CC01] Error updating device state: TypeError: Cannot read properties of undefined (reading 'alias')
    at file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/homekitPowerStrip.ts:37:81
    at processTicksAndRejections (node:internal/process/task_queues:105:5)
```

**Comment by ZeliardM** [comment]
@plasticfist Ok, I'll keep poking around. I suppose it all have to do with the sheer number of devices and the way I am setting everything up. Need to account for more devices.

**Comment by ZeliardM** [comment]
@plasticfist give beta.15 a try and let me know the outcome

**Comment by plasticfist** [comment]
```
[12/30/2024, 7:38:13 PM] Homebridge v1.8.5 (HAP v0.12.3) (Homebridge 800D) is running on port 51124.
[12/30/2024, 7:38:13 PM] 

NOTICE TO USERS AND PLUGIN DEVELOPERS
> Homebridge 2.0 is on the way and brings some breaking changes to existing plugins.
> Please visit the following link to learn more about the changes and how to prepare:
> https://github.com/homebridge/homebridge/wiki/Updating-To-Homebridge-v2.0

[12/30/2024, 7:38:13 PM] [KasaPython] homebridge-kasa-python v2.7.0-beta.15, node v22.12.0, homebridge v1.8.5, api v2.7 Initializing...
[12/30/2024, 7:38:13 PM] [KasaPython] Version file updated to version 2.7.0-beta.15
[12/30/2024, 7:38:13 PM] [KasaPython] Plugin version changed, virtual python environment will be recreated.
[12/30/2024, 7:38:34 PM] [KasaPython] Discovering devices...
[12/30/2024, 7:38:42 PM] [KasaPython] Adding HomeKit device: [Upstairs bug zapper SP5] IOT.SMARTPLUGSWITCH [8006A5F7EA0A58346642B5171A7F07AA20DDFE8F] at host [10.0.1.104]
...
... adds all devices. then after a pause
...
[12/30/2024, 7:39:17 PM] [KasaPython] Exception during getSysInfo post request: Unable to query the device 10.0.1.137:9999: [Errno 104] Connection reset by peer
[12/30/2024, 7:39:17 PM] [KasaPython] [Power Strip CB17] Error updating device state: TypeError: Cannot read properties of undefined (reading 'alias')
    at file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/homekitPowerStrip.ts:37:81
    at processTicksAndRejections (node:internal/process/task_queues:105:5)
[12/30/2024, 7:39:27 PM] [KasaPython] Exception during getSysInfo post request: Unable to query the device 10.0.1.33:9999: [Errno 104] Connection reset by peer
[12/30/2024, 7:39:27 PM] [KasaPython] [Power Strip DAA0] Error updating device state: TypeError: Cannot read properties of undefined (reading 'alias')
    at file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/homekitPowerStrip.ts:37:81
    at processTicksAndRejections (node:internal/process/task_queues:105:5)
(node:105862) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 periodicDeviceDiscoveryComplete listeners added to [EventEmitter]. MaxListeners is 10. Use emitter.setMaxListeners() to increase limit
(Use `node --trace-warnings ...` to show where the warning was created)
[12/30/2024, 7:40:03 PM] [KasaPython] [Power Strip CB17] Device is offline or platform is shutting down, cannot get value for characteristic On
[12/30/2024, 7:40:03 PM] [KasaPython] [Power Strip CB17] Device is offline or platform is shutting down, cannot get value for characteristic OutletInUse
[12/30/2024, 7:40:03 PM] [KasaPython] [Power Strip CB17] Device is offline or platform is shutting down, cannot get value for characteristic On
[12/30/2024, 7:40:03 PM] [KasaPython] [Power Strip CB17] Device is offline or platform is shutting down, cannot get value for characteristic OutletInUse
[12/30/2024, 7:40:03 PM] [KasaPython] [Power Strip CB17] Device is offline or platform is shutting down, cannot get value for characteristic On
[12/30/2024, 7:40:03 PM] [KasaPython] [Power Strip CB17] Device is offline or platform is shutting down, cannot get value for characteristic OutletInUse
[12/30/2024, 7:40:03 PM] [KasaPython] [Power Strip CB17] Device is offline or platform is shutting down, cannot get value for characteristic On
[12/30/2024, 7:40:03 PM] [KasaPython] [Power Strip CB17] Device is offline or platform is shutting down, cannot get value for characteristic OutletInUse
...
... and more of the above for various devices then
...
[12/30/2024, 7:40:22 PM] [KasaPython] [Power Strip DAA0] Device is offline or platform is shutting down, cannot get value for characteristic On
[12/30/2024, 7:40:22 PM] [KasaPython] [Power Strip DAA0] Device is offline or platform is shutting down, cannot get value for characteristic OutletInUse
[12/30/2024, 7:41:37 PM] [KasaPython] Exception during getSysInfo post request: Unable to query the device 10.0.1.35:9999: [Errno 104] Connection reset by peer
[12/30/2024, 7:41:37 PM] [KasaPython] [Power Strip 2480] Error updating device state: TypeError: Cannot read properties of undefined (reading 'alias')
    at file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/homekitPowerStrip.ts:37:81
    at processTicksAndRejections (node:internal/process/task_queues:105:5)
[12/30/2024, 7:42:03 PM] [KasaPython] Exception during getSysInfo post request: Unable to query the device 10.0.1.223:9999: [Errno 104] Connection reset by peer
[12/30/2024, 7:42:03 PM] [KasaPython] [Power Strip CF00] Error updating device state: TypeError: Cannot read properties of undefined (reading 'alias')
    at file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/homekitPowerStrip.ts:37:81
    at processTicksAndRejections (node:internal/process/task_queues:105:5)
[12/30/2024, 7:42:05 PM] [KasaPython] [Power Strip CF00] Error updating device state: TypeError: Cannot read properties of undefined (reading 'children')
    at file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/homekitPowerStrip.ts:215:68
    at Array.forEach (<anonymous>)
    at file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/homekitPowerStrip.ts:211:44
    at processTicksAndRejections (node:internal/process/task_queues:105:5)
    at HomeKitDevicePowerStrip.updateState (file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/homekitPowerStrip.ts:237:5)
    at Timeout.<anonymous> (file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/homekitPowerStrip.ts:260:9)
[12/30/2024, 7:42:07 PM] [KasaPython] Exception during getSysInfo post request: Unable to query the device 10.0.1.106:9999: [Errno 104] Connection reset by peer
[12/30/2024, 7:42:07 PM] [KasaPython] [Power Strip C7EF] Error updating device state: TypeError: Cannot read properties of undefined (reading 'alias')
    at file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/homekitPowerStrip.ts:37:81
    at processTicksAndRejections (node:internal/process/task_queues:105:5)
```

**Comment by ZeliardM** [comment]
@plasticfist let's give beta.16 a shot, thanks for all the back and forth. You seem to have the most devices out of anyone who has reported anything so far and I want to work this out.

**Comment by ZeliardM** [comment]
@plasticfist if you can, enable debug mode in the Homebridge UI and post those logs, it will have much more information for me.

**Comment by plasticfist** [comment]
in debug mode, we can see that max listener count stays at listener count +10, climbing to 42, but after errors occur, it changes to max 10 and then throws the listener error.   When not it debug mode we don't see it climbing to 42 first.  The rest of the log is just verbose successful discovery messages (noise).
```
[12/31/2024, 8:22:57 PM] [KasaPython] periodicDeviceDiscoveryEmitter periodicDeviceDiscoveryComplete listener count: 32
[12/31/2024, 8:22:57 PM] [KasaPython] periodicDeviceDiscoveryEmitter max listener count: 42
[12/31/2024, 8:22:57 PM] [KasaPython] Device discovery completed
[12/31/2024, 8:22:57 PM] [KasaPython] Setting up periodic device discovery
[12/31/2024, 8:22:57 PM] [KasaPython] Periodic device discovery setup completed
...
[12/31/2024, 8:38:10 PM] [KasaPython] Exception during getSysInfo post request: Unable to query the device 10.0.1.27:9999: [Errno 104] Connection reset by peer
[12/31/2024, 8:38:10 PM] [KasaPython] [Power Strip 9379] Error updating device state: TypeError: Cannot read properties of undefined (reading 'alias')
    at file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/homekitPowerStrip.ts:37:81
    at processTicksAndRejections (node:internal/process/task_queues:105:5)
[12/31/2024, 8:38:10 PM] [KasaPython] [Power Strip 9379] Error updating device state: TypeError: Cannot read properties of undefined (reading 'alias')
    at file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/homekitPowerStrip.ts:37:81
    at processTicksAndRejections (node:internal/process/task_queues:105:5)
...
(node:134048) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 periodicDeviceDiscoveryComplete listeners added to [EventEmitter]. MaxListeners is 10. Use emitter.setMaxListeners() to increase limit
```

Here is the initiallization debug for reference (before the above), just in case it is useful
```
[12/31/2024, 8:22:39 PM] Homebridge v1.8.5 (HAP v0.12.3) (Homebridge 800D) is running on port 51124.
[12/31/2024, 8:22:39 PM] 

NOTICE TO USERS AND PLUGIN DEVELOPERS
> Homebridge 2.0 is on the way and brings some breaking changes to existing plugins.
> Please visit the following link to learn more about the changes and how to prepare:
> https://github.com/homebridge/homebridge/wiki/Updating-To-Homebridge-v2.0

[12/31/2024, 8:22:39 PM] [KasaPython] homebridge-kasa-python v2.7.0-beta.16, node v22.12.0, homebridge v1.8.5, api v2.7 Initializing...
[12/31/2024, 8:22:39 PM] [KasaPython] Verifying environment
[12/31/2024, 8:22:39 PM] [KasaPython] Checking Node.js version
[12/31/2024, 8:22:39 PM] [KasaPython] Node.js version v22.12.0 satisfies the requirement ^18.20.5 || ^20.18.1 || ^22.12.0 || ^23.5.0
[12/31/2024, 8:22:39 PM] [KasaPython] Checking Homebridge version
[12/31/2024, 8:22:39 PM] [KasaPython] Homebridge version 1.8.5 satisfies the requirement >= 1.8.4
[12/31/2024, 8:22:39 PM] [KasaPython] Checking for upgrade at path: /var/lib/homebridge/kasa-python/kasa-python-version.json
[12/31/2024, 8:22:39 PM] [KasaPython] Stored version: 2.7.0-beta.16
[12/31/2024, 8:22:39 PM] [KasaPython] No upgrade needed, version is up to date.
[12/31/2024, 8:22:39 PM] [KasaPython] Finished launching
[12/31/2024, 8:22:39 PM] [KasaPython] Checking Python environment
[12/31/2024, 8:22:39 PM] [KasaPython] Running PythonChecker with isUpgrade: false
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Starting python environment check...
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Ensuring plugin directory exists: /var/lib/homebridge/kasa-python
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Plugin directory already exists: /var/lib/homebridge/kasa-python
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Ensuring system Python version is supported
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Getting system Python versions
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Running command: python --version
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command PID: undefined
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Running command: python3 --version
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command PID: 134060
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Running command: python3.11 --version
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command PID: 134061
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Running command: python3.12 --version
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command PID: undefined
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Running command: python3.13 --version
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command PID: undefined
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command closed with exit code: -2
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command closed with exit code: -2
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command closed with exit code: 0
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command finished.
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command closed with exit code: 0
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command finished.
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Found Python versions: [ '3.11.2' ]
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Using Python executable: python3.11
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] System Python version is supported
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Ensuring virtual environment is created
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Virtual environment exists: true
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Virtual environment already exists
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Ensuring virtual environment uses correct Python home
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Getting Python home for executable: /var/lib/homebridge/kasa-python/.venv/bin/python3.11
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Running command: /var/lib/homebridge/kasa-python/.venv/bin/python3.11 /var/lib/homebridge/node_modules/homebridge-kasa-python/dist/python/pythonHome.py
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command PID: 134064
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command closed with exit code: -2
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command closed with exit code: 0
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command finished.
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Getting Python home for executable: python3.11
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Running command: python3.11 /var/lib/homebridge/node_modules/homebridge-kasa-python/dist/python/pythonHome.py
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command PID: 134065
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command closed with exit code: 0
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command finished.
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Virtual environment Python home: /usr/bin
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] System Python home: /usr/bin
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Python homes match
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Ensuring virtual environment pip is up to date
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Getting virtual environment pip version
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Running command: /var/lib/homebridge/kasa-python/.venv/bin/pip3.11 --version
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command PID: 134066
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command closed with exit code: 0
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command finished.
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Virtual environment pip version: 24.3.1
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Fetching most recent pip version from PyPI
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Most recent pip version fetched: 24.3.1
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Virtual environment pip is up to date
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Checking if virtual environment requirements are satisfied
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Running command: /var/lib/homebridge/kasa-python/.venv/bin/pip3.11 freeze
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command PID: 134067
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command closed with exit code: 0
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Command finished.
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Current virtual environment packages: {
  "aiofiles": "24.1.0",
  "aiohappyeyeballs": "2.4.4",
  "aiohttp": "3.11.11",
  "aiosignal": "1.3.2",
  "anyio": "4.7.0",
  "asyncclick": "8.1.7.2",
  "attrs": "24.3.0",
  "blinker": "1.9.0",
  "cffi": "1.17.1",
  "click": "8.1.8",
  "cryptography": "44.0.0",
  "flask": "3.1.0",
  "frozenlist": "1.5.0",
  "h11": "0.14.0",
  "h2": "4.1.0",
  "hpack": "4.0.0",
  "hypercorn": "0.17.3",
  "hyperframe": "6.0.1",
  "idna": "3.10",
  "itsdangerous": "2.2.0",
  "jinja2": "3.1.5",
  "markupsafe": "3.0.2",
  "mashumaro": "3.15",
  "multidict": "6.1.0",
  "priority": "2.0.0",
  "propcache": "0.2.1",
  "pycparser": "2.22",
  "python-kasa": "0.9.0",
  "quart": "0.20.0",
  "sniffio": "1.3.1",
  "typing-extensions": "4.12.2",
  "uvicorn": "0.34.0",
  "werkzeug": "3.1.3",
  "wsproto": "1.2.0",
  "yarl": "1.18.3"
}
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Required packages: {
  "python-kasa": "0.9.0",
  "quart": "0.20.0",
  "uvicorn": "0.34.0"
}
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Requirements satisfied: true
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Virtual environment requirements are already satisfied
[12/31/2024, 8:22:39 PM] [KasaPython] [Python Check] Python environment check completed successfully
[12/31/2024, 8:22:39 PM] [KasaPython] Getting available port
[12/31/2024, 8:22:39 PM] [KasaPython] Port assigned: 45477
[12/31/2024, 8:22:39 PM] [KasaPython] Initializing DeviceManager
[12/31/2024, 8:22:39 PM] [KasaPython] DeviceManager initialized
[12/31/2024, 8:22:39 PM] [KasaPython] Starting Kasa API with script: /var/lib/homebridge/node_modules/homebridge-kasa-python/dist/python/startKasaApi.py
[12/31/2024, 8:22:39 PM] [KasaPython] Running command: /var/lib/homebridge/kasa-python/.venv/bin/python3 /var/lib/homebridge/node_modules/homebridge-kasa-python/dist/python/startKasaApi.py 45477
[12/31/2024, 8:22:39 PM] [KasaPython] Command PID: 134068
[12/31/2024, 8:22:39 PM] [KasaPython] Command started and returning process.
[12/31/2024, 8:22:40 PM] [KasaPython] Process data received.
[12/31/2024, 8:22:40 PM] [KasaPython] Kasa API process started successfully
[12/31/2024, 8:22:40 PM] [KasaPython] Waiting for server at http://127.0.0.1:45477/health with timeout 30000ms and interval 1000ms
[12/31/2024, 8:22:40 PM] [KasaPython] Server not responding yet, retrying...
[12/31/2024, 8:22:41 PM] [KasaPython] Server not responding yet, retrying...
[12/31/2024, 8:22:42 PM] [KasaPython] Server not responding yet, retrying...
[12/31/2024, 8:22:43 PM] [KasaPython] Server responded successfully
[12/31/2024, 8:22:43 PM] [KasaPython] Discovering devices...
[12/31/2024, 8:22:57 PM] [KasaPython] Number of devices discovered: 32
[12/31/2024, 8:22:57 PM] [KasaPython] Processing discovered device: 8006A5F7EA0A58346642B5171A7F07AA20DDFE8F
[12/31/2024, 8:22:57 PM] [KasaPython] periodicDeviceDiscoveryEmitter periodicDeviceDiscoveryComplete listener count: 0
[12/31/2024, 8:22:57 PM] [KasaPython] periodicDeviceDiscoveryEmitter max listener count: 10
[12/31/2024, 8:22:57 PM] [KasaPython] Adding HomeKit device: [Upstairs bug zapper SP5] IOT.SMARTPLUGSWITCH [8006A5F7EA0A58346642B5171A7F07AA20DDFE8F] at host [10.0.1.104]
[12/31/2024, 8:22:57 PM] [KasaPython] Creating HomeKit device for: {
  alias: 'Upstairs bug zapper SP5',
  child_num: 0,
  device_id: '8006A5F7EA0A58346642B5171A7F07AA20DDFE8F',
  device_type: 'IOT.SMARTPLUGSWITCH',
  host: '10.0.1.104',
  hw_ver: '1.0',
  mac: '9C:53:22:55:8D:3E',
  model: 'EP25',
  state: false,
  sw_ver: '1.0.14'
}
[12/31/2024, 8:22:57 PM] [KasaPython] HomeKit device is a Plug: EP25
[12/31/2024, 8:22:57 PM] [KasaPython] [Upstairs bug zapper SP5] Existing Platform Accessory found [8006A5F7EA0A58346642B5171A7F07AA20DDFE8F] [d31a29d1-8e7f-400c-94ef-675d5712c0a6] category: OUTLET
[12/31/2024, 8:22:57 PM] [KasaPython] [Upstairs bug zapper SP5] Correcting Platform Accessory lastSeen from: Tue Dec 31 2024 20:22:38 GMT+0000 (Coordinated Universal Time) to: Tue Dec 31 2024 20:22:57 GMT+0000 (Coordinated Universal Time)
[12/31/2024, 8:22:57 PM] [KasaPython] [Upstairs bug zapper SP5] Initializing HomeKitDevicePlug for device: Upstairs bug zapper SP5
[12/31/2024, 8:22:57 PM] [KasaPython] [Upstairs bug zapper SP5] Starting polling for device: Upstairs bug zapper SP5
[12/31/2024, 8:22:57 PM] [KasaPython] HomeKit device [Upstairs bug zapper SP5] IOT.SMARTPLUGSWITCH [8006A5F7EA0A58346642B5171A7F07AA20DDFE8F] successfully added
[12/31/2024, 8:22:57 PM] [KasaPython] periodicDeviceDiscoveryEmitter periodicDeviceDiscoveryComplete listener count: 1
[12/31/2024, 8:22:57 PM] [KasaPython] periodicDeviceDiscoveryEmitter max listener count: 11
[12/31/2024, 8:22:57 PM] [KasaPython] Processing discovered device: 8006812C5AC692EB6F92FB2DE765DE1820271487
[12/31/2024, 8:22:57 PM] [KasaPython] periodicDeviceDiscoveryEmitter periodicDeviceDiscoveryComplete listener count: 1
[12/31/2024, 8:22:57 PM] [KasaPython] periodicDeviceDiscoveryEmitter max listener count: 11
[12/31/2024, 8:22:57 PM] [KasaPython] Adding HomeKit device: [Power Strip C7EF]...
```

**Comment by ZeliardM** [comment]
@plasticfist I appreciate the logs. Would you be able to attach the whole plugin log file so I can trace all the logs, I am getting the errors from you, but I'm not sure what is going on before or after.

Would you also be able to add --trace-warnings to the Node Options under the Settings in the Web UI to see if we can more details about what is going on.

Thanks!

**Comment by ZeliardM** [comment]
@plasticfist Give beta.18 a shot and let me know. Changed to a task queue to try and keep from overloading the devices and a reconnect feature and some more logging for the kasaApi.py

If you can, enable debug mode before you run it and grab the full logs from the plug-in, please and thank you!

**Comment by ZeliardM** [comment]
@plasticfist beta.19 fixes a few more problems from the previous changes so if you can use that one instead. Again, enable advanced python logging in the plug-in settings, debug mode in the homebridge web ui settings, add --trace-warnings to the node options in the homebridge web ui settings, restart and post the full log capturing the issues.

**Comment by BrentWilliams76** [comment]
@ZeliardM, since I am having similar problems, I thought I'd give you my logs.  These are using beta.19 and all of the debug settings you requested above.  Hope you find what you are looking for!

[homebridge.log.txt](https://github.com/user-attachments/files/18291777/homebridge.log.txt)

**Comment by ZeliardM** [comment]
@BrentWilliams76 @plasticfist Ok, I think I see where it's coming from when periodic discovery is being done. I believe it's not emptying the device cache and then when it tries to get sys_info, it's trying to use the old connection instead of the new one from the periodic discovery. I will try to work out some better logic here for the device cache and see what I can come up with. Thanks!

**Comment by ZeliardM** [comment]
@plasticfist @BrentWilliams76 Ok, let's give beta.20 a shot and let me know. If you want when pulling the log, go to the plugins page and click the dots next to the Kasa Python Plug-in and click Plug-In Logs and it will only put information in relevant to the plug-in when you post it. Let it run for two periodic device discovery phases, so a little over 10 minutes should give a good indication for me. Thanks!

**Comment by BrentWilliams76** [comment]
Been running for 15 minutes now.  I've seen three discoveries and zero errors or warnings.  I did not have the advanced debugging on, but can if you still want or need logs.

**Comment by ZeliardM** [comment]
If you wouldn't mind, I would like to see and make sure things are looking good from all ends. Also, how is the response time when turning devices on and off dimming and such?

**Comment by BrentWilliams76** [comment]
Here you go.  I did see a few STDERR messages scroll by, so maybe something of interest still there for you.

And response time seems to still be very good.

[homebridge-kasa-python.log.txt](https://github.com/user-attachments/files/18294031/homebridge-kasa-python.log.txt)

**Comment by ZeliardM** [comment]
@plasticfist @BrentWilliams76 give beta.21 a shot. If you can post another log of everything after 2 or 3 periodic discovery cycles, that would be good as well.

It seems to be functioning better, but I added in some additional closing of connections that aren't being used. The errors that are still showing for the Query failing is a retriable exception in the python-kasa library and it will automatically retry the connection.

I added in some waiting for the device polling to keep from calling the devices too quickly.

**Comment by BrentWilliams76** [comment]
Here's a log from beta.21
[homebridge-kasa-python.log (1).txt](https://github.com/user-attachments/files/18302596/homebridge-kasa-python.log.1.txt)

**Comment by ZeliardM** [comment]
@plasticfist @BrentWilliams76 can you give beta.22 a shot and let me know the outcome?

**Comment by BrentWilliams76** [comment]
Updated to beta.22 and was immediately getting errors after device discovery.  Log attached.  Switching back to beta.21 now since wife is back home from holiday and Homebridge is "production critical" again.
[homebridge-kasa-python.log (2).txt](https://github.com/user-attachments/files/18417401/homebridge-kasa-python.log.2.txt)

**Comment by ZeliardM** [comment]
Yea, I have errors with the config changes, will work on that later tonight and get beta 23 out and let everyone know. Thanks!

**Comment by ZeliardM** [comment]
give beta.23 a shot

**Comment by BrentWilliams76** [comment]
Updated to beta.24 and got similar errors to what I received in beta.22.  The log is attached. 

[homebridge-kasa-python.log.txt](https://github.com/user-attachments/files/18425672/homebridge-kasa-python.log.txt)

**Comment by ZeliardM** [comment]
@plasticfist @BrentWilliams76 please give beta.25 a try and let me know.

**Comment by BrentWilliams76** [comment]
So far, so good with beta.25.  Here's a log that includes two subsequent device discovery cycles.

Thanks for your work on this!

[homebridge-kasa-python.log (1).txt](https://github.com/user-attachments/files/18455556/homebridge-kasa-python.log.1.txt)

**Comment by ZeliardM** [comment]
@BrentWilliams76 would you be able to enable Homebridge Debug Mode and the Plug-In Advanced Python Logging to get more detail on the Unclosed client connections?

**Comment by BrentWilliams76** [comment]
Does this give you what you want to see?

[homebridge-kasa-python.log (2).txt](https://github.com/user-attachments/files/18456190/homebridge-kasa-python.log.2.txt)

**Comment by plasticfist** [comment]
I'm seeing all the same type of errors as @BrentWilliams76, and with one more, see below

```
[1/17/2025, 3:25:55 PM] [KasaPython] Discovered 32 devices
[1/17/2025, 3:27:55 PM] [Homebridge UI] Accessory layout changes saved for freego.
[1/17/2025, 3:28:00 PM] [KasaPython] Error during getSysInfo post request: TypeError: Cannot use 'in' operator to search for 'sys_info' in undefined
    at DeviceManager.updateDeviceAlias (file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/deviceManager.ts:40:31)
    at DeviceManager.getSysInfo (file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/deviceManager.ts:181:12)
    at processTicksAndRejections (node:internal/process/task_queues:105:5)
    at file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/homekitSwitch.ts:37:36
[1/17/2025, 3:28:00 PM] [KasaPython] [Office light] Error updating device state: TypeError: Cannot read properties of undefined (reading 'alias')
    at file:///var/lib/homebridge/node_modules/homebridge-kasa-python/src/devices/homekitSwitch.ts:38:81
    at processTicksAndRejections (node:internal/process/task_queues:105:5)
```

**Comment by BrentWilliams76** [comment]
One thing I can note is that the plugin seems to be consuming a lot of CPU.  Granted, I am running Homebridge on an older machine.  (It's an HP Microserver Gen7 with AMD Turion II Neo N40L CPU) but I'm still a bit surprised to see the plugin consistently consume ~35% of the CPU capacity.  Prior to beta.25, I was seeing similarly high CPU utilization with regular spikes up to ~95-100% CPU every few seconds and then back down to the fairly consistent baseline level (~50%, of which ~35% is kasa-python).  I don't recall seeing that high CPU load prior to jumping on the latest beta train, but I also cannot say that it wasn't there.

I'm not sure if this is known or expected, just wanted to point it out.  I have been considering moving to a somewhat newer and more powerful server platform (HP Micro Gen8 with Intel Xeon E3-1220 v2), but still old by today's standards.  Since the massive peaks have disappeared, that move is not urgent, but I might make time to do it anyway.

**Comment by ZeliardM** [comment]
I have seen mine hovering around 15%, when breaking everything up into multiple processes it lowers it, but when the calls and connections come in they go to different processes that devices are connected to causing multiple connections to devices which cause a lot of the other issues. Still trying to balance everything correctly and make it more effecient.

**Comment by ZeliardM** [comment]
I also have a lot fewer devices, so that may also play into it as well.

**Comment by BrentWilliams76** [comment]
I just got a few error messages hit and it caused the overall CPU load to increase to ~77%.  (Possible that something else has caused the CPU load... just coincidental timing?)  Sorry... I had turned off advanced logging hoping for some CPU reduction (don't think it helped) but means I don't have all the details in the logs.  Nevertheless, here's what I captured.

[homebridge-kasa-python.log (3).txt](https://github.com/user-attachments/files/18459670/homebridge-kasa-python.log.3.txt)

**Comment by BrentWilliams76** [comment]
OK... nevermind on the ADDITIONAL CPU load.  Looks like it was an hourly backup job running for one of my Macs.  CPU load is back down to ~50% overall, including ~35% kasa-python.

**Comment by ZeliardM** [comment]
@plasticfist @BrentWilliams76 Give beta.26 a shot and let me know if there are any issues.

**Comment by BrentWilliams76** [comment]
beta.26 installed.  No issues seen yet.  Log file attached.

[homebridge-kasa-python.log (3).txt](https://github.com/user-attachments/files/18470576/homebridge-kasa-python.log.3.txt)

**Comment by BrentWilliams76** [comment]
I've noticed with at least the last several beta versions (and continuing with beta.26) some errors related to two of my smart plugs.  Here's a snippet of the logs reflecting these errors.  Full log is attached.  It is probably helpful for you to know that the affected Smart Plugs are model HS103, Hardware rev 2.1, Firmware 1.1.4.  I have other smart plugs (HS103, HW 5.0, FW 1.0.13 and EP10, HW 1.0, FW 1.0.5) that do not seem to be affected, so I'm sure it's these older units not behaving the same.

[1/20/2025, 9:46:40 AM] [KasaPython] STDERR: GetSysInfo failed: cannot access local variable 'device' where it is not associated with a value
GetSysInfo route error: cannot access local variable 'device' where it is not associated with a value
[1/20/2025, 9:46:40 AM] [KasaPython] Exception during getSysInfo post request: cannot access local variable 'device' where it is not associated with a value
[1/20/2025, 9:46:40 AM] [KasaPython] [Office Desk Lamp] Error updating device state: TypeError: Cannot read properties of undefined (reading 'alias')
    at file:///volume1/homebridge/node_modules/homebridge-kasa-python/src/devices/homekitPlug.ts:36:81
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
[1/20/2025, 9:49:32 AM] [KasaPython] [Office Desk Lamp] Device is offline or platform is shutting down, cannot get value for characteristic On
[1/20/2025, 9:49:32 AM] [KasaPython] [Office Desk Lamp] Device is offline or platform is shutting down, cannot get value for characteristic OutletInUse

[homebridge-kasa-python.log (4).txt](https://github.com/user-attachments/files/18479786/homebridge-kasa-python.log.4.txt)

**Comment by ZeliardM** [comment]
@BrentWilliams76 Ok, I'll take a look. Thanks!

**Comment by ZeliardM** [comment]
@BrentWilliams76 Can you enable the Advanced Python Logging and Debug mode so I can see more of what is going on and where it's losing the device connection? I have it narrowed down, but I don't know which piece is where it can't find the device.

**Comment by BrentWilliams76** [comment]
Naturally, when I am looking for the problem, I don't see it.  But I'll try to keep the advanced logging and debug mode active.  I also see beta.28 out, so upgrading there now.

**Comment by ZeliardM** [comment]
@plasticfist @BrentWilliams76 Can y'all give beta.36 a shot and let me know how everything goes?

**Comment by BrentWilliams76** [comment]
Running beta.36 now.  No errors noticed in first discovery cycle.

**Comment by plasticfist** [comment]
FYI.  updated, here are some notes for reference, but I  don't have time these days to do debug logging, etc. sorry
- errors were being thrown on two devices, a motion sensor wall switch and a 6 outlet strip (these are not unique I have others).  I reset the wall switch and removed the strip from power for 15 seconds.  both started working.  Note:  I can't ever recall having to do this when using Kasa app but caveats apply...  before the reset the outlet strip was working with kasa app, but the wall switch was not even on the network anymore, it was in a bad state...
- After reseting them, and restarting homebridge, pushing the accessory button in homebridge often seems to do nothing (homebridge bug maybe?) nothing in log, nothing happens
- restart homebridge again, buttons seem to work, for a short while, then stop working (note, using the homebridge accessory buttons)
- updates about on/off TO homebridge seem to usually work (see note below), but have about a 3 second delay
- homebridge and/or plugin seems to often decide it can't turn on/off kasa accessories.  No errors in log, not even any messages to log about trying to turn on/off accessory...  
- Note:  device discovery it seems to blocks use of kasa accessories from homebridge, can't say this is the only cause of non-functional kasa accessories, but the plugin seems to be in device discovery mode... a lot...

**Comment by ZeliardM** [comment]
@plasticfist thanks for the update.
- I don't do anything with motion sensors in the plug-in, I may look into it, if it's something you'd like to see. I had seen some issues with several of the betas when I was going back and forth on a lot of things, that some of the devices locked up and had to be manually power cycled before they would work again, I'm hopeful that on the latest release that it no longer happens.
- I have had issues with the buttons in Homebridge as well for other plug-ins, so I am not sure about that. I will be honest, I don't use the Homebridge interface, once I have the bridge in HomeKit, I use everything there.
- I'm glad the delay is down; I've done my best to keep the connections to the devices open and hopefully the slight delay is better for your use.
- The plug-in will mark devices offline if it is having trouble communicating with them, if the communication can be restored in the next periodic update, it should resolve the issue, but it could come down to the number of devices and the amount of communication required on the network for keeping a constant polling of the devices. The errors and such will show up with the Advanced Python Logging/Debug Logging.
- Periodic Device Discovery will block communication to the devices as it is trying to update and get new communication paths to the devices and update their information, if you try to change it during a discovery, it will not work through Homebridge or the HomeKit App, but the Kasa/Tapo App will still allow those changes and if the device is updated during a discovery by the native app, the plug-in should reflect that change now as well.

Please give the full release v2.7.0 a shot and let me know if any of these problems are now resolved.

**Comment by ZeliardM** [comment]
@BrentWilliams76 I have pushed a full release on this to v2.7.0, let me know if you have any further problems.

**Comment by plasticfist** [comment]
@ZeliardM I'll try to remember to test from homekit only (iphone), it is much more responsive, no failures in my quick test just now, while the homebridge UI buttons were still randomly unresponsive (at that same moment in time...)
EXCEPT, when the plugin was in discovery mode, then the homekit eventually turned the light on (10-15 seconds? later)
a feature request would be to allow existing accessories to continue working during a discovery update, complicates the code possibly requiring a background process I know.  (e.g. like the game of life, next cycle is in a new matrix)

**Comment by ZeliardM** [comment]
@plasticfist I'll look into it and see if it's something I can make happen and we can push out some betas and test. Thanks for all your feedback and help in making this better for everyone!

**Comment by ZeliardM** [comment]
@BrentWilliams76 @plasticfist can you give v2.7.1-beta.1 a shot and let me know if things are looking more like you'd prefer and if there are any errors or problems showing up with the debug/advanced python logging going on? Thanks!

**Comment by ZeliardM** [comment]
@plasticfist One thing I am noticing is that motion sensors may be more difficult to implement since I am polling the device for status updates and if motion is missed during the polling, it may not look or act correctly. I will have to think on this quite a bit more and see if there is a way that I can either get it to constantly poll just the pir sensor in the backend and then push those updates to the homebridge plug-in, will see what I can come up with as I go along and let you know. They just did some big updates to the pir sensors in the older switches, so that got some good updates this last release from python-kasa.

**Comment by BrentWilliams76** [comment]
> [@BrentWilliams76](https://github.com/BrentWilliams76) [@plasticfist](https://github.com/plasticfist) can you give v2.7.1-beta.1 a shot and let me know if things are looking more like you'd prefer and if there are any errors or problems showing up with the debug/advanced python logging going on? Thanks!

Been running for 24 hours and I have not seen any errors in my environment.

**Comment by ZeliardM** [comment]
Excellent, thanks sir!

**Comment by plasticfist** [comment]
v2.7.1-beta.1 ran fine from release until I just upgraded to latest (some prior versions would eventually get into a bad state where nothing worked claiming that all devices were not responding, so more time on task will help give confidence that issue is gone for sure)

Currently "Discovering devices..." takes almost exactly 1 minute, during which all devices are unresponsive.

**Comment by ZeliardM** [comment]
@plasticfist That's weird. All that was done is merge from the beta release to the main, nothing else. Can you restart homebridge and see if that makes a difference?

**Comment by plasticfist** [comment]
discovery process time required varies a little, 50 to at times 60+ seconds

2.7.1
```
[1/30/2025, 4:12:04 PM] [KasaPython] Discovering devices...
[1/30/2025, 4:12:54 PM] [KasaPython] Discovered 33 devices
[1/30/2025, 4:17:04 PM] [KasaPython] Discovering devices...
[1/30/2025, 4:17:55 PM] [KasaPython] Discovered 33 devices
...
[1/30/2025, 4:42:12 PM] [KasaPython] Discovering devices...
[1/30/2025, 4:43:16 PM] [KasaPython] Discovered 33 devices
```

**Comment by ZeliardM** [comment]
Is there any way you can enable debugging with the advanced python logging so I can see where it might be hanging up?

**Comment by plasticfist** [comment]
I think it is an issue with plugin not failing fast when a device requires authentication? (why this one is asking for it, is a separate issue likely I have to figure out at my end)   I tried adding it to home assistant and it prompted for auth, so...

only one msg/error, this is the second? "discovering devices", the first only reports "adding device..." messages, and no "Discovered N devices", during that first step there were no errors, but then after

```
[1/30/2025, 4:51:11 PM] [KasaPython] Discovering devices...
[1/30/2025, 4:51:11 PM] [KasaPython] STDERR: AuthenticationError: 10.0.1.98
[1/30/2025, 4:52:06 PM] [KasaPython] Discovered 33 devices
[1/30/2025, 4:56:11 PM] [KasaPython] Discovering devices...
[1/30/2025, 4:56:11 PM] [KasaPython] STDERR: AuthenticationError: 10.0.1.98
[1/30/2025, 4:57:08 PM] [KasaPython] Discovered 33 devices
```

This is an EP2, can control it from Kasa, I checked, it says it is on the latest firmware, specific version is
Firmware version 1.0.3 Build 240529, Rel.145252
Hardware Version 2.6
very likely I have at least two of these.  So I need to figure out why this one is behaving differently (though both might be doing this and the other is unplugged...).  At one point I got a two pack that say they can integrate with homekit, likely it's one of those...

**Comment by ZeliardM** [comment]
Yea, if the username and password are not provided, it will timeout after a default timeout of I believe 5 seconds for responses from devices and with a large amount of devices, could take a while for it to go through the whole process.

Are you saying that during the first discovery after the plug-in starts that it does not give this error? But during the periodic discovery, it does? And it does not say the number of devices discovered?

If the native device filtering is on, it will not include them in the plug-in but they would still be discovered since it does a network wide discovery. Without doing discovery by device ip individually listed in manual devices, if it's on the network it will get picked up and error out the authentication no real way around that.

**Comment by plasticfist** [comment]
> Yea, if the username and password are not provided, it will timeout after a default timeout of I believe 5 seconds for responses from devices and with a large amount of devices, could take a while for it to go through the whole process.
> 
> Are you saying that during the first discovery after the plug-in starts that it does not give this error? But during the periodic discovery, it does? And it does not say the number of devices discovered?

Correct

>  timeout after a default timeout of I believe 5 seconds

NOTE: there is as far as I can tell from error messages only 1 device that does the auth error, so I think that means the auth timeout issue only adds 5 seconds total (in the 50-60 seconds discovery functionality pause)

**Comment by ZeliardM** [comment]
Correct, and if you have forty or so devices the timing for everything really does add up about a second or so per device. I could have it process each device individually and send on instead of all of them at once so it doesn't have to wait for every device to send on. I could look into that.

**Comment by plasticfist** [comment]
separating the discovery process from the active accessories list is likely the best (and maybe only) way to eliminate the blocking problem, various approaches may be possible to do it either for the whole list (e.g. background) with a lock and swap at the end, or at the individual device level (e.g. lock and update individual), perhaps these ideas might require
a separate thread/process (maybe not), but for the rare collision some lock/wait logic is probably wise, etc.  Yeah, I understand that this would be more complicated, and for just a handful of devices a lot of unneeded complexity.

from an unbounded python perspective, I would have thought 33 devices (or even a 1000) would be reasonable to poll in parallel, however I haven't looked at the code and do not know the homekit plugin architecture, so probably there are things I don't know about the limitations that the plugin has to live inside... and/or low minimum host requirements, etc.
Also adds complexity.

**Comment by ZeliardM** [comment]
The discovery is done through a UDP broadcast with the python-kasa library. When credentials are provided for devices that require them, the responses are usually fast, but I have seen it slow down significantly when they are not provided and the timeout ensues.

I will look at possibly increasing the speed on my plug-in side of things and seeing if devices could be handled independently. Proceeding with the discovery for the plug-in while the python-kasa discovery is ongoing. There are on_discovered callbacks that can be used so I can see how that would work for each device to process through on_discovered.

Will let you know. Thanks!

**Comment by ZeliardM** [comment]
@plasticfist Still working on a beta for handling the devices as they are discovered to see if we can speed things up some. Will keep you in the loop. Thanks!

**Comment by ZeliardM** [comment]
@plasticfist I am working on a new beta for 2.8.0 that should implement a streaming aspect for device discovery making it a little more fluid. Once I have something to test, I'll let you know!

**Comment by ZeliardM** [comment]
@plasticfist @BrentWilliams76 give beta v2.8.0-beta.0 a shot and let me know.

**Comment by BrentWilliams76** [comment]
Upgraded and everything seems good after first discovery.  Will update if I notice any changes.

**Comment by ZeliardM** [comment]
@plasticfist How does the speed on v2.8.0-beta.1 look for you?

**Comment by ZeliardM** [comment]
@plasticfist try v2.8.0-beta.3 and let me know.
